.gitmodules

```diff
@@ -124,3 +124,6 @@
 [submodule "hw/mcu/mindmotion/mm32sdk"]
 	path = hw/mcu/mindmotion/mm32sdk
 	url = https://github.com/zhangslice/mm32sdk.git
+[submodule "hw/mcu/gd/nuclei-sdk"]
+	path = hw/mcu/gd/nuclei-sdk
+	url = https://github.com/Nuclei-Software/nuclei-sdk.git
```

Comment:
I completely agree that the nuclei-sdk is overkill for this port, but I had chosen it over the multiple other repositories of the GD32VF firmware on github because their repository is the most up to date one and actively maintained by nucleisys which provided the RV32 core IP. Furthermore there is a bug in the startup code that throws an compilation error if `-Wundef` and `-Werror` is used with GCC, for which I had [opened a PR.](https://github.com/Nuclei-Software/nuclei-sdk/pull/30) Therefore I would vote to keep this despite being more then necessary, what do you think?

The other repositories:

https://github.com/riscv-mcu/GD32VF103_Firmware_Library (Semi-official, contains a note to use nuclei-sdk, last activity in October 2020)

https://github.com/GigaDevice-Semiconductor/GD32VF103_Firmware_Library (Is this official?, last real activity in November 2020)

Reply:
I pull the nuclei and did a quick check, the https://github.com/GigaDevice-Semiconductor/GD32VF103_Firmware_Library is more updated in term of the mcu driver. The last activities you said in 2020 is update the mcu lib to v1.1. while nuclei still uses the v1.0 version. nuclei is more maintained by it is mostly for their platform, tinyusb is only interested in low level mcu driver (clock, register def). Therefore I think it make more sense to move to the said repo.

Reply:
I'm a bit confused the latest firmware library from GigaDevice is 1.1.2 from June 2021 but can only be found on http://www.gd32mcu.com/en/download/0?kw=GD32VF1 so all the repos are outdated. Judging from the activity in repositories all but nuclei-sdk seem to be unmaintained, as no pull request are merged or answered.

I hope that I don't bother you with this discussion, but I think that nuclei-sdk is the better choice from a maintenance perspective and I would happily open a PR to update the supplied firmware in nuclei-sdk. If that is not an option at all, I'll switch to the suggested https://github.com/GigaDevice-Semiconductor/GD32VF103_Firmware_Library repository.

Reply:
I am not bothered at all, I have looked a bit more closer. To be honest, at first I thought Nuclei is a 3rd party, but look like they are IP creator similar to ARM. And Giga put Nuclei's N200 risc-v core into their silicon. Furthermore, Nuclei seems to spend much more effort on the software/firmware side than Giga does. And I could see why you prefer to go with their sdk. Feel free to correct me if I get it wrong. 

https://github.com/Nuclei-Software/NMSIS is indeed a really great repo which is set to do what CMSIS does for ARM. I think we could actually depend on this nmsis. However, I still don't want to force any other users to use nuclei sdk just to run the usb stack. They may already be on another platform/framework, that could cause an conflict. Unless there is something that we would really need from sdk that we couldn't write it ourself.

To sum up, do you think it is good enough to just add 2 submodules
- semi official https://github.com/riscv-mcu/GD32VF103_Firmware_Library for low level mcu, I don't mind that it is old v1.0, mcu driver doesn't change much. And USB does not use much of them. 
- https://github.com/Nuclei-Software/NMSIS for providing abstraction for risc-v core for eclic enabled if needed.

This is actually very similar to stm32 with an cmisis and driver combo, https://github.com/hathach/tinyusb/blob/master/hw/bsp/stm32f103bluepill/board.mk#L2

Note: when integrating tinyusb to an application, you can freely use nuclei, and have your own board setup, just changes the include path to the corresponding. The idea is if we use an smaller set of files (mcu lib), it would be more portable.

Reply:
> I am not bothered at all, I have looked a bit more closer. 

That is good to hear!

You are totally right about Nuclei, they provided the N200/BumbleBee Core for the GD32VF103 to GigaDevice and kind of co-developed the firmware library. I'm not sure why GigaDevice doesn't show a bit more effort with their firmware, it is a bit frustrating.

> I still don't want to force any other users to use nuclei sdk just to run the usb stack. They may already be on another platform/framework, that could cause an conflict. Unless there is something that we would really need from sdk that we couldn't write it ourself.

I'm totally with you that we don't want conflicts in other projects. I think you are referring to namespace conflicts like multiple definitions of the same symbol but correct me here. My point is that the parts that we include and actively use in the compilation of tinyusb from `nuclei-sdk` is really just the NMSIS library and the gd32vf103 firmware library with modifications from nuclei to play along nicely with NMSIS (see below). The other header and source files related to different socs are not included in the build, it is just the bare minimum. So the conflicts should not increase by using `nuclei-sdk` and should be the same as using the separate NMSIS and GD32VF103 firmware repositories.

On the other hand I had started to change the current port to use the two independent repositories as suggested and ran into trouble. The NMSIS library wants specific [defines](https://github.com/Nuclei-Software/NMSIS/blob/master/Device/Nuclei/NUCLEI_N/Include/NUCLEI_N.h#L120-L142) that describe the capabilities of the GD32VF103. Unfortunately we can't just use this file because the GD32VF103 doesn't have a FPU and has other IRQs defined. So we would have to provide this block ourselfs in a separate header file before [including nmsis_core.h in the dcd driver](https://github.com/hathach/tinyusb/blob/3eb54d878af0492a85fab2811b5f84187c2f703c/src/portable/st/synopsys/dcd_synopsys.c#L98-L99). 

But by doing so we have a board specific header-file in a rather generic driver, which is not good. The `nuclei-sdk` provides these defines in [gd32vf103.h](https://github.com/Nuclei-Software/nuclei-sdk/blob/master/SoC/gd32vf103/Common/Include/gd32vf103.h) which is different from the [gd32vf103.h](https://github.com/riscv-mcu/GD32VF103_Firmware_Library/blob/master/Firmware/GD32VF103_standard_peripheral/gd32vf103.h) that is supplied with the firmware library. And they actually modified the firmware library in other parts to play a long nicely with NMSIS framework. The GD32VF103 firmware library is really just meant to be used on its own. There maybe even more conflicts that but I haven't explored that way further. 

Given the above reasons and addressing your concerns about the conflicts could a dependency on just `nuclei-sdk` be the better option? What do you think?

Reply:
> That is good to hear!

You know this much better than me, there is obvious reason why you prefer the nuclei sdk
 
> You are totally right about Nuclei, they provided the N200/BumbleBee Core for the GD32VF103 to GigaDevice and kind of co-developed the firmware library. I'm not sure why GigaDevice doesn't show a bit more effort with their firmware, it is a bit frustrating.

Yeah, I figured it out, Giga doesn't spend much effort like ST do with their stm32

> But by doing so we have a board specific header-file in a rather generic driver, which is not good. The `nuclei-sdk` provides these defines in [gd32vf103.h](https://github.com/Nuclei-Software/nuclei-sdk/blob/master/SoC/gd32vf103/Common/Include/gd32vf103.h) which is different from the [gd32vf103.h](https://github.com/riscv-mcu/GD32VF103_Firmware_Library/blob/master/Firmware/GD32VF103_standard_peripheral/gd32vf103.h) that is supplied with the firmware library. And they actually modified the firmware library in other parts to play a long nicely with NMSIS framework. The GD32VF103 firmware library is really just meant to be used on its own. There maybe even more conflicts that but I haven't explored that way further.
> 
> Given the above reasons and addressing your concerns about the conflicts could a dependency on just `nuclei-sdk` be the better option? What do you think?

My ultimate goal is not having usb stack to depend on nuclei-sdk to run. User should just compile it with any GD32VF103_Firmware_Library variant, especially one downloaded from the Giga site since that is what user will do. With that in mind, in the fact the nuclei has to modify the library to get it work with nmsis, I think we actually have to change the approach a bit.
1. We can keep using nuclei-sdk, and use it for running example only ie. bsp folder.
2. However we shouldn't use any of nuclei-sdk and/or nmsis within the `src` folder, i.e the `dcd_synopsys.c`. The reason is it will force user to use the modified firmware lib from nuclei and not be able to use **official** lib downloaded from Giga website. Even it is not well maintained, it is still what most user would use (unfortunately).

Reply:
That sounds good to me! I'll change the implementation to meet these constraints, maybe the core code in src can even be implemented without any reliance on any gd32vf103 firmawre library or nmsis/nuclei-sdk.

Reply:
Thanks for patiently following my suggestion, yeah, the example bsp is mostly for demonstrating/testing and maintaining. If nuclei-sdk make it easier for us, no problem at all, I wasn't sure before and kind of worry it will become an dependency due to the nmsis. If we could get it dcd_synopsys neutral then that would be no problem.

Reply:
Perfect! I have cleaned up and fixed a bug with the systick timer not beeing reloaded. I decided to go with `nuclei-sdk` inside bsp for obvious reasons ;-). Let me know what you think!